architecture/device_tree.rst

```diff
@@ -62,15 +89,46 @@ by early boot and passed to non-secure world are the following:
 
     - Add a PSCI description node if none found.
 
-Early boot DTB located in non-secure memory can be accessed by OP-TEE core only
-during its initialization, before non-secure world boots.
+Early boot DTB can be accessed by OP-TEE core only during its initialization,
+before non-secure world boots as it is expected the DTB memory location has
+lileky been replaced with runtime contextes content.
+
+Assuming there is no embedded DTB (section :ref:`embedded_dtb`) OP-TEE core
+discovers the main memory address ranges from the non-secure DTB.
+
+.. _external_dtb_overlay:
+
+Early boot device tree overlay
+******************************
+There are two possibilities for OP-TEE core to provide a device tree
+overlay to the non-secure world.
+
+    - Append OP-TEE nodes to an existing DTB overlay located in early boot DTB.
+      (``CFG_DT_ADDR`` or boot argument register ``R2``/``X2``).
+
+    - Generate a new DTB overlay image at location defined by ``CFG_DT_ADDR``.
+
+In the later case, memory referred by configuration directive ``CFG_DT_ADDR``
+shall not contain a valid DTB image when OP-TEE core is booted. A subsequent
+non-secure boot stage should merge the OP-TEE DTB overlay image into
+another DTB.
+
+A typical bootflow for this would be Arm-Trusted-Firmawre -> OP-TEE -> U-Boot
+with U-Boot merging the OP-TEE DTB overlay located at ``CFG_DT_ADDR``
+into a DTB U-Boot has loaded from elsewhere.
+
+This functionality is enabled when ``CFG_EXTERNAL_DTB_OVERLAY=y``.
+
+.. _embedded_dtb:
 
 Embedded Secure Device Tree
 ***************************
-When OP-TEE core is built with configuration directive ``CFG_EMBED_DTB=y``
+When OP-TEE core is built with configuration directive ``CFG_EMBED_DTB=y``,
 directive ``CFG_EMBED_DTB_SOURCE_FILE`` shall provide the relative path of the
 DTS file inside directory ``core/arch/$(ARCH)/dts`` from which a DTB is
 generated and embedded in a read-only section of OP-TEE core.
 
-In this case the device tree address passed to the OP-TEE entry point by the
-bootloader is ignored, only the embedded device tree is accessible.
+Refer to ``core/include/kernel/dt.h`` for API to access embedded DTB.
+
+Section :ref:`generic_boot_and_dtb` documents the generic boot sequence
```

Comment:
s/generic_boot_and_dtb/generic_boot_and_dtbs/ (this is what Travis complains about).